components/eamxx/src/diagnostics/pbl_entrainment_budget.cpp

```diff
@@ -0,0 +1,332 @@
+#include "diagnostics/pbl_entrainment_budget.hpp"
+
+#include "diagnostics/pbl_entrainment_budget_util.hpp"
+#include "ekat/ekat_workspace.hpp"
+#include "ekat/kokkos/ekat_kokkos_utils.hpp"
+#include "share/util/scream_universal_constants.hpp"
+
+namespace scream {
+
+PBLEntrainmentBudget::PBLEntrainmentBudget(const ekat::Comm &comm,
+                                           const ekat::ParameterList &params)
+    : AtmosphereDiagnostic(comm, params) {
+  // Nothing to do here
+}
+
+void PBLEntrainmentBudget::set_grids(
+    const std::shared_ptr<const GridsManager> grids_manager) {
+  using namespace ekat::units;
+  using namespace ShortFieldTagsNames;
+
+  auto grid             = grids_manager->get_grid("Physics");
+  const auto &grid_name = grid->name();
+
+  const auto nondim = Units::nondimensional();
+
+  // Set the index map and units map
+  PBLEntrainmentBudgetDiagUtil eadu;
+  m_index_map = eadu.index_map;
+  m_units_map = eadu.units_map;
+  m_ndiag     = eadu.size;
+
+  if(eadu.pblinvalg == "temperature-inversion") {
+    m_pblinvalg = 1;
+  } else if(eadu.pblinvalg == "thetal-only") {
+    m_pblinvalg = 2;
+  } else if(eadu.pblinvalg == "qt_only") {
+    m_pblinvalg = 3;
+  } else {
+    EKAT_ERROR_MSG(
+        "Error! Invalid pblinvalg. Only temperature-inversion, thetal-only, "
+        "and "
+        "qt_only are currently supported.\n");
+  }
+
+  // Ensure m_index_map and m_units_map match
+  EKAT_REQUIRE_MSG(
+      m_index_map.size() == m_units_map.size(),
+      "Error! Some inconsistency in PBLEntrainmentBudget: index and units "
+      "maps do not match!\n");
+  // Ensure m_index_map and m_ndiag match
+  EKAT_REQUIRE_MSG(
+      static_cast<int>(m_index_map.size()) == m_ndiag,
+      "Error! Some inconsistency in PBLEntrainmentBudget: m_ndiag and index "
+      "map do not match!\n");
+
+  m_ncols = grid->get_num_local_dofs();
+  m_nlevs = grid->get_num_vertical_levels();
+
+  // Define layouts we need (both inputs and outputs)
+  FieldLayout scalar2d_layout{{COL, LEV}, {m_ncols, m_nlevs}};
+  FieldLayout vector1d_layout{{COL, CMP}, {m_ncols, m_ndiag}};
+
+  // The fields required for this diagnostic to be computed
+  // Get qc and qv
+  add_field<Required>("qc", scalar2d_layout, kg / kg, grid_name);
+  add_field<Required>("qv", scalar2d_layout, kg / kg, grid_name);
+  // Get T_mid, p_mid
+  add_field<Required>("T_mid", scalar2d_layout, K, grid_name);
+  add_field<Required>("p_mid", scalar2d_layout, Pa, grid_name);
+  // Get pseudo_density
+  add_field<Required>("pseudo_density", scalar2d_layout, Pa, grid_name);
+  // Get radiation up and down terms
+  add_field<Required>("SW_flux_dn", scalar2d_layout, W / m * m, grid_name);
```

Comment:
Watch out, these units here will resolve to `W`, since you don't have parentheses!

Note: this will have _zero_ impact on this code working, since I/O does not check the units of the requested fields. But we should fix this, a) to avoid confusion and b) in case I/O starts to check units of the diags required fields.